File: duna-core/src/main/java/io/duna/core/inject/RemoteServiceBinderModule.java
/*
 * Copyright (c) 2016 Duna Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.inject;

import com.google.inject.AbstractModule;
import com.google.inject.UnsafeTypeLiteral;
import io.duna.core.context.ClasspathScanner;

import java.lang.ref.WeakReference;
import java.util.logging.Level;
import java.util.logging.LogManager;
import java.util.logging.Logger;

public class RemoteServiceBinderModule extends AbstractModule {

    private static final Logger logger = LogManager
        .getLogManager().getLogger(RemoteServiceBinderModule.class.getName());

    private final WeakReference<ClasspathScanner> classpathScanner;

    public RemoteServiceBinderModule(ClasspathScanner classpathScanner) {
        this.classpathScanner = new WeakReference<>(classpathScanner);
    }

    @Override
    protected void configure() {
        logger.info(() -> "");

        ClasspathScanner localClasspathScanner = classpathScanner.get();
        if (localClasspathScanner == null) localClasspathScanner = new ClasspathScanner();

        final ClasspathScanner effectiveScanner = localClasspathScanner;

        effectiveScanner
            .getRemoteServices()
            .parallelStream()
            .forEach(contractClassName -> {
                Class<?> contractClass;

                try {
                    contractClass = Class.forName(contractClassName, false, this.getClass().getClassLoader());
                } catch (ClassNotFoundException e) {
                    logger.log(Level.WARNING, e, () -> "Error while trying to bind contract to service proxy.");
                    return;
                }

                if (!contractClass.isInterface()) {
                    logger.warning(() -> "Unable to bind " + contractClass.getName() + ". " +
                        "Contracts must be declared as interfaces.");
                    return;
                }

                UnsafeTypeLiteral contractTypeLiteral = new UnsafeTypeLiteral(contractClass);


            });
    }
}

File: duna-core/src/main/kotlin/io/duna/core/inject/RemoteServiceBinderModule.kt
/*
 * Copyright (c) 2016 Duna Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.inject

import com.google.inject.AbstractModule
import com.google.inject.UnsafeTypeLiteral
import io.duna.core.classpath.ClasspathScanner
import io.duna.core.proxy.RemoteServiceProxyFactory
import java.lang.reflect.Modifier
import java.util.logging.LogManager

internal object RemoteServiceBinderModule : AbstractModule() {

  @JvmStatic
  private val logger = LogManager.getLogManager()
    .getLogger(RemoteServiceBinderModule::class.java.name)

  private val classLoader = RemoteServiceProxyFactory()

  /**
   * TODO Support qualifiers for proxies
   */
  override fun configure() {
    logger.info { "Binding proxies for remote services" }

    val remoteContracts = ClasspathScanner.getRemoteServices()
        .map { Class.forName(it) }

    remoteContracts.forEach contractForEach@ { contractClass ->
      if (!contractClass.isInterface && !Modifier.isAbstract(contractClass.modifiers)) {
        logger.warning { "Unable to bind ${contractClass.canonicalName}. " +
          "Contracts must be either an interface or abstract class." }
        return@contractForEach
      }

      val contractTypeLiteral = UnsafeTypeLiteral(contractClass)
      val proxyClass = classLoader.loadProxyForService(contractClass).newInstance()

      logger.info { "Bound proxy for ${contractClass.canonicalName}" }

      requestInjection(proxyClass)
      bind(contractTypeLiteral).toInstance(proxyClass)
    }

    logger.info { "Proxies for remote services bound" }
  }
}
